presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java

```diff
@@ -48,98 +42,266 @@ public ArrayIntersectFunction(@TypeParameter("E") Type elementType)
         pageBuilder = new PageBuilder(ImmutableList.of(elementType));
     }
 
-    private static IntComparator intBlockCompare(Type type, Block block)
-    {
-        return new AbstractIntComparator()
-        {
-            @Override
-            public int compare(int left, int right)
-            {
-                if (block.isNull(left) && block.isNull(right)) {
-                    return 0;
-                }
-                if (block.isNull(left)) {
-                    return -1;
-                }
-                if (block.isNull(right)) {
-                    return 1;
-                }
-                return type.compareTo(block, left, block, right);
-            }
-        };
-    }
-
     @TypeParameter("E")
     @SqlType("array(E)")
     public Block intersect(
-            @OperatorDependency(operator = LESS_THAN, returnType = StandardTypes.BOOLEAN, argumentTypes = {"E", "E"}) MethodHandle lessThanFunction,
             @TypeParameter("E") Type type,
             @SqlType("array(E)") Block leftArray,
             @SqlType("array(E)") Block rightArray)
     {
+        if (leftArray.getPositionCount() < rightArray.getPositionCount()) {
+            Block tempArray = leftArray;
+            leftArray = rightArray;
+            rightArray = tempArray;
+        }
+
         int leftPositionCount = leftArray.getPositionCount();
         int rightPositionCount = rightArray.getPositionCount();
 
-        if (leftPositionCount == 0) {
-            return leftArray;
-        }
         if (rightPositionCount == 0) {
             return rightArray;
         }
 
-        if (leftPositions.length < leftPositionCount) {
-            leftPositions = new int[leftPositionCount];
+        if (pageBuilder.isFull()) {
+            pageBuilder.reset();
         }
 
-        if (rightPositions.length < rightPositionCount) {
-            rightPositions = new int[rightPositionCount];
+        BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(0);
+
+        TypedSet rightTypedSet = new TypedSet(type, rightPositionCount, "array_intersect");
+        for (int i = 0; i < rightPositionCount; i++) {
+            rightTypedSet.add(rightArray, i);
+        }
+
+        // The intersected set can have at most rightPositionCount elements
+        TypedSet intersectTypedSet = new TypedSet(type, rightPositionCount, "array_intersect");
+        int intersectCount = 0;
+        for (int i = 0; i < leftPositionCount; i++) {
+            if (rightTypedSet.contains(leftArray, i) && !intersectTypedSet.contains(leftArray, i)) {
+                intersectTypedSet.add(leftArray, i);
+                type.appendTo(leftArray, i, blockBuilder);
+                intersectCount++;
+            }
+        }
+
+        return blockBuilder.getRegion(blockBuilder.getPositionCount() - intersectCount, intersectCount);
```

Comment:
I believe that `intersectTypedSet` already contains the output data, so maybe modify `TypedSet` to return `elementBlock.build`.